File: zul/src/org/zkoss/zul/Panelchildren.java
/* Panelchildren.java

	Purpose:
		
	Description:
		
	History:
		Jun 10, 2008 11:47:15 AM , Created by jumperchen

Copyright (C) 2008 Potix Corporation. All Rights Reserved.

{{IS_RIGHT
	This program is distributed under LGPL Version 3.0 in the hope that
	it will be useful, but WITHOUT ANY WARRANTY.
}}IS_RIGHT
*/
package org.zkoss.zul;

import java.io.IOException;

import org.zkoss.zk.ui.Component;
import org.zkoss.zk.ui.UiException;
import org.zkoss.zk.ui.sys.ContentRenderer;
import org.zkoss.zul.impl.XulElement;

/**
 * Panelchildren is used for {@link Panel} component to manage each child who will
 * be shown in the body of Panel.
 * Note that the size of Panelchildren is automatically calculated by Panel so both
 * {@link #setWidth(String)} and {@link #setHeight(String)} are read-only.
 * 
 * <p>Default {@link #getZclass}: z-panel-children.
 * 
 * @author jumperchen
 * @since 3.5.0
 */
public class Panelchildren extends XulElement {
	
	public void beforeParentChanged(Component parent) {
		if (parent != null && !(parent instanceof Panel))
			throw new UiException("Wrong parent: "+parent);
		super.beforeParentChanged(parent);
	}
	
	/**
	 * This method is unsupported. Please use {@link Panel#setWidth(String)} instead.
	 */
	public void setWidth(String width) {
		throw new UnsupportedOperationException("readonly");
	}
	/**
	 * This method is unsupported. Please use {@link Panel#setHeight(String)} instead.
	 */
	public void setHeight(String height) {
		throw new UnsupportedOperationException("readonly");
	}
	public String getHflex() {
		Component parent = getParent();
		return parent == null ? null : ((Panel) parent).getHflex();
	}
	public String getVflex() {
		Component parent = getParent();
		return parent == null ? null : ((Panel) parent).getVflex();
	}
	/**
	 * This method is unsupported. Please use {@link Panel#setHflex(String)} instead.
	 * @since 5.5.0
	 */
	public void setHflex(String flex) {
		throw new UnsupportedOperationException("readonly");
	}
	/**
	 * This method is unsupported. Please use {@link Panel#setVflex(String)} instead.
	 * @since 5.5.0
	 */
	public void setVflex(String flex) {
		throw new UnsupportedOperationException("readonly");
	}
	
	protected void smartUpdate(String attr, Object value) {
		super.smartUpdate(attr, value); // provides a bridge so it can be called by Panel
	}
	protected void renderProperties(ContentRenderer renderer) throws IOException {
		super.renderProperties(renderer);
		render(renderer, "hflex", getHflex());
		render(renderer, "vflex", getVflex());
	}

	// super
	public String getZclass() {
		return _zclass == null ?  "z-panel-children" : _zclass;
	}
}

File: zul/src/main/resources/web/js/zul/wnd/Panelchildren.ts
/* Panelchildren.ts

	Purpose:

	Description:

	History:
		Mon Jan 12 18:31:03     2009, Created by jumperchen

Copyright (C) 2008 Potix Corporation. All Rights Reserved.

This program is distributed under LGPL Version 2.1 in the hope that
it will be useful, but WITHOUT ANY WARRANTY.
*/
/**
 * Panelchildren is used for {@link zul.wnd.Panel} component to manage each
 * child who will be shown in the body of Panel.
 * Note that the size of Panelchildren is automatically calculated by Panel so both
 * {@link setWidth} and {@link setHeight} are read-only.
 *
 * @defaultValue {@link getZclass}: z-panelchildren.
 */
@zk.WrapClass('zul.wnd.Panelchildren')
export class Panelchildren extends zul.ContainerWidget {
	override parent?: zul.wnd.Panel;

	/**
	 * This method is unsupported. Please use {@link zul.wnd.Panel.setHeight} instead.
	 */
	override setHeight(height?: string): this {
		// Empty on purpose. Shoudn't do anything.
		return this;
	}

	/**
	 * This method is unsupported. Please use {@link zul.wnd.Panel.setWidth} instead.
	 */
	override setWidth(width?: string): this { // readonly
		// Empty on purpose. Shoudn't do anything.
		return this;
	}

	// super
	/** @internal */
	override domClass_(no?: zk.DomClassOptions): string {
		const out: string[] = [],
			scls = super.domClass_(no);
		if (scls)
			out.push(scls);
		if (!no?.zclass) {
			const zcls = this.getZclass(),
				parent = this.parent!;
			if (!parent.getTitle() && !parent.caption)
				out.push(zcls + '-noheader');
			if (!parent._bordered())
				out.push(zcls + '-noborder');
		}
		return out.join(' ');
	}

	/** @internal */
	override updateDomStyle_(): void {
		super.updateDomStyle_();
		if (this.desktop)
			zUtl.fireSized(this.parent!);
	}
}